### ISSUE TITLE
Remove huge JavaScript parsers from Prettier core to reduce package size

### ISSUE THREAD

**Issue by sosukesuzuki** [prettier/prettier#13912]
https://github.com/prettier/prettier/issues/13818#issuecomment-1308082760

## Overview

Prettier package includes too many JavaScript parsers. And it makes the size of the npm package much larger than it needs to be.

So, I propose the change to keep only the limited parsers that are really needed in the core, and provide and maintain the rest as external parser plugins.

## How

Only `babel`, `babel-flow`, and `babel-ts` remain in the Prettier core. Other parsers `typescript`, `acorn`, `espree`, `meriyah`, and `flow` are provided as external plugins(with names like `@prettier/plugin-parser-typescript` for example).

This will also make it easier to introduce new parsers. For example, we could provide a hermes and SWC as a JavaScript parser for Prettier.

### 1. monorepo

With this change we will have to publish multiple packages(I thinks parsers packages version should be synchronized with the core). 

Also we need to share common postprocess and test cases for JS parsers.

So I think we can introduce monorepo by Yarn workspace like below:

```
.
│   ├── packages
│   │   ├── prettier
│   │   ├── prettier-plugin-parser-typescript
│   │   ├── prettier-plugin-parser-meriyah
│   │   ├── prettier-plugin-parser-acorn
│   │   ├── prettier-plugin-parser-flow
│   │   └── prettier-plugin-parser-espree
│   ├── internals
│   │   ├── eslint-plugin-prettier-internal
│   │   ├── util-js-parser-postprocess
│   │   └── test-cases-js-parsers
│   └── apps
│   │   └── website
```

(There may be other things that should be in the internal package that I forgot.)

**This is a difficult change to revert and should be discussed thoroughly.**

### 2. multiple repositories

Just maintain and develop each parser as a separate repository

e.g. `github.com/prettier/plugin-parser-typescript`, `github.com/prettier/plugin-parser-meriyah`, etc

---

What do you think?

**Comment by fisker** [comment]
I prefer separate Node.js version and standard version first, because all parsers(`plugins/*.mjs`) I may use someday. But I don't think I'll ever use the UMD version.

BTW, the package name should be

```diff
- @prettier/plugin-parser-typescript
+ @prettier/plugin-typescript
```

**Comment by sosukesuzuki** [comment]
What is standard version?

> BTW, the package name should be

The name `@prettier/plugin-typescript` makes it sound like it even includes the print algorithm. In fact, only the parser is included there.

**Comment by fisker** [comment]
standalone , sorry.

**Comment by fisker** [comment]
> only the parser is included there.

Printer is not required to be a plugin, and printer can be included if we publish plugins separately.

**Comment by fisker** [comment]
For the file structure, I prefer

```
.
│   ├── packages
│   │   ├── core (will publish as `@prettier/core`)
│   │   ├── cli (will publish as `@prettier/cli`)
│   │   ├── document (will publish as `@prettier/doc` or `@prettier/document`)
│   │   ├── js 
│   │   │   ├── typescript  (will publish as `@prettier/plugin-typescript`)
│   │   │   ├── meriyah  (will publish as `@prettier/plugin-meriyah`)
│   │   │   ├── ...  (will publish as `@prettier/plugin-...`)
│   │   │   └── tests  (to test all js plugins)
│   │   ├── markdown  (will publish as `@prettier/plugin-markdown`)
│   │   ├── css  (will publish as `@prettier/plugin-css`)
│   │   ├── yaml  (will publish as `@prettier/plugin-yaml`)
│   │   └── ...other plugins
```

Except for the js plugins, they can even still stay in `src/language-xxx/`, nobody says package has to be inside `packages/`, add a `package.json` in those dirs will work.

**Comment by fisker** [comment]
Make our codebase monorepo may also break "install-from-github" feature, I don't think `workspace:` in dependency will work.

**Comment by sosukesuzuki** [comment]
With the structure you showed, users will have to use `@prettier/core` and `@prettier/cli` when using Prettier? Ideally, when using Prettier, users would just need to install `prettier`.

**Comment by fisker** [comment]
No, we'll add them to `prettier` package's dependencies. But they can install them separately.

**Comment by sosukesuzuki** [comment]
Okay, if we provide packages in the style of Jest, no problem.

**Comment by liuxingbaoyu** [comment]
> Make our codebase monorepo may also break "install-from-github" feature, I don't think workspace: in dependency will work.

Just a personal opinion: if we go ahead with https://github.com/prettier/prettier/issues/6286 in the future, this will have the same impact.
At the same time, with the increasing number of complex projects in the JS world, "install-from-github" may not be so important anymore.
